### ISSUE TITLE
[Onboarding] Present on-boarding experience/routes when first opening app

### ISSUE THREAD

**Issue by conrad-vanl** [ApollosProject/apollos-prototype#519]
### User Story

As a user I should be presented the `Auth` route then the `Onboarding` route every time I log in.

### Description

Everytime a user isn't logged in, we should force the `Auth` route. After login, we should then direct the user to the `Onboarding` route. After `Onboarding`, we can take the user to the home screen.

Since we persist your login (we currently reset this each build but will be changing that) you should only have to login ~1 time per device. Every time you login on a new device, you need to be presented with the onboarding experience again to enable notifications and location services. For MVP, we are comfortable just re-showing onboarding every time you login. Again, we currently reset your auth state with every build, and we will change that to not function that way in a separate task.

### Acceptance Criteria

* [ ] It should show the `Auth` route if you're not logged in
* [ ] After logging in, push/switch to the `Onboard` route
* [ ] After onboarding, go back to the home tab

### Possible Implementation & Open Questions

* The yet-to-be-merge [`ui auth` package](https://github.com/ApollosProject/apollos-prototype/pull/591) adds an `afterAuth` style callback to make some of this easier, but I think we can go ahead and wire this up with the current `Auth` route in `apolloschurchapp` until that's merged
* We probably want to maintain separate `Auth` and `Onboarding` routes on the root navigator

### Other Resources

* Basecamp: https://3.basecamp.com/3926363/buckets/6947926/todolists/1506896432

**Comment by conrad-vanl** [comment]
@richarddubay said (on Slack):

> the `Auth` component/route doesn't have any logic to tell whether or not someone is already signed in. It looks like that logic exists on the `Connect` component/route. Should we default the route to `Connect` then? Or update the `Auth` route to have the same `getLoginState` query that the `Connect` route does? Or maybe something else

**Comment by conrad-vanl** [comment]
So I'm not 100% sure why you need the "login state" logic inside of `Auth`.

I think what you probably need to do is somewhere near the root-level of our app stack, check for user login state, and if user is not logged in, redirect to the `Auth` component. This root-level can use a `<Query />` component querying for login state. Something like:

```
const getLoginState = gql`
  query {
    isLoggedIn @client
  }
`;

<Query query={getLoginState}>
  {({ data: { isLoggedIn } = {}, loading: innerLoading }) => {
    if (isLoggedIn) showLoginSomehow();
  })}
</Query>
```

**Comment by conrad-vanl** [comment]
Could also follow something similar to React-Navigation's recommendations here: https://reactnavigation.org/docs/en/auth-flow.html

**Comment by richarddubay** [comment]
Question: When trying to navigate to one of two routes based on the return value from a query, have you ever experienced both routes being navigated to as the result of the query changes?

I'm checking for `isLoggedIn` and the app is supposed to navigate to one route when the result is `true` and a different one when it is `false`.

What I'm experiencing is that when I'm logged in, the query returns false first, and it brings up the false route, but then the query returns true and it then brings up the true route.

Is there any way to wait on the final query result before navigating? FWIW, I'm checking that `loading` is false before I do anything.

Also, any ideas on why it would render both routes? Shouldn't it pick one and go to it?

**Comment by richarddubay** [comment]
Also, in attempting to try the React Navigation idea, I tried just pulling the `authToken` from `AsyncStorage` but that doesn't seem to get cleared when you log out. So that didn't work either.

**Comment by conrad-vanl** [comment]
@richarddubay I was thinking through these questions yesterday. So I think your first issue is that apollo queries hit a few different states - initial data, loading, done loading. It will probably always mount with a false state, then update to a true state. I'm not quite sure about the inconsistent loading situation. You might try logging the additional arguments to see what's going on / changing with each render.

If you don't want the dual route scenario to be possible, you probably need to do something more like the react-navigation example, where you don't push routes, but instead you render three possible states: a loading/unknown state, the logged in state, and the logged out state.

I'm not sure what the `AsyncStorage` bit is in the react navigation idea, but I would still use a Query with `isLoggedIn` query, the only part from the react navigation example I was looking at taking was how they use a `SwitchNavigator` with the three states. Initially show a loading component that's responsible for directing you to the right screen.

**Comment by conrad-vanl** [comment]
Another way I've seen it done is by not using a navigator at all responsible for switching states ....

Just have a component that uses the logged in query, and based on the value either renders an `AuthNavigator` component or a `AppNavigator` component.

**Comment by richarddubay** [comment]
Maybe I'm wrong here ... could very well be ... but I'm not sure that using the `Query` component is the way to go here. Doesn't seem to matter what I do, if the value of `isLoggedIn` can change after `loading` is no longer true, there is no way to escape having the app display both routes. If I had to guess, the reason that the React Navigation example works is because they are just reading a value from storage and not depending on the result of a changing query. 

Here is my [branch](https://github.com/ApollosProject/apollos-prototype/tree/519-routing) currently with one of many options I've tried if anyone wants to play with it or see if they see where I'm going wrong. Thanks all.
